views-core/src/main/java/io/micronaut/views/AsyncViewsRenderer.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2017-2019 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.views;
+
+import io.micronaut.core.io.Writable;
+import io.micronaut.http.HttpRequest;
+import io.micronaut.http.MutableHttpResponse;
+import org.reactivestreams.Publisher;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+
+/**
+ * Asynchronous rendering interface for views in Micronaut. This interface works with reactive types to allow the event
+ * loop to take over when the renderer is paused, in cases where renderers support such signals.
+ *
+ * @see ViewsRenderer for the synchronous version
+ * @author Sam Gammon
+ * @since 1.3.0
+ */
+public interface AsyncViewsRenderer extends BaseViewsRenderer {
+  /**
+   * @param viewName view name to be render
+   * @param data     response body to render it with a view
+   * @param request  HTTP request
+   * @param response HTTP response object.
+   * @return A writable where the view will be written to.
+   */
+  @Nonnull
+  Publisher<MutableHttpResponse<Writable>> render(
```

Comment:
The only thing I am unsure about is the current synchronous version returns a `Writable` whilst this returns a response. Also I am wondering whether instead the `Publisher` should be `Publisher<ByteBuffer>` since what if the view rendering supports outputting the view in chunks?

Reply:
@graemerocher this has been my concern precisely - i haven't quite figured out how to make the buffering here make sense.

publishing as `ByteBuffer` cleans that up significantly, I think. i'll give it a try

Reply:
@graemerocher - I'm working on converting it, but it appears the `HttpServerFilter` interface expects a `Publisher<MutableHttpResponse<?>>`. do you think it would be okay to change this interface, and/or introduce a new one that worked with `ByteBuffer`/`ByteBuf`? Or is that even necessary?

I was thinking of going with `ByteBuf` because I presume views are generally a server-side activity, so the `netty-buffer` dependency is probably fine. If you have any objection to this let me know

Reply:
@graemerocher i could wrap each chunked `ByteBuf`/`ByteBuffer` in a `MutableHttpResponse`, but I don't know if Micronaut will chunk that, or expect just one `MutableHttpResponse`.

Reply:
@sgammon Sorry for the late replay, been travelling recently. We can't change any core interfaces as this would be a breaking change.

IMO the signature should be:

```java
  @NonNull Flowable<ByteBuffer> render(
    @Nonnull String viewName,
    @Nullable Object data);
}
```

Where `ByteBuffer` is a `io.micronaut.core.io.buffer.ByteBuffer`. This more closely mirrors the current blocking signature variation. It probably should be named `ReactiveViewRenderer` instead of `AsyncViewRenderer` as well.

If you return a `Flowable<ByteBuffer>` then Micronaut will deal with chunking the responses and sending them back.

Reply:
hey there @graemerocher,

i've run into an issue with the `ReactiveViewRenderer`. essentially, `ViewsFilter` is implemented as an `HttpServerFilter`, so its interface is `Publisher<? extends HttpResponse<?>> doFilter(...);`.

so when you say it should return a `Flowable<ByteBuffer>`, do you actually mean `Flowable<HttpResponse<ByteBuffer>>`? it does seem to work that way but i can't find the chunking logic in micronaut.